### PR TITLE
multi: Decouple orphan handling from blockchain.

### PR DESCRIPTION
**Pull request by davecgh**
This decouples and removes the orphan handling from blockchain in favor of implementing it in the block manager as part of the overall effort to decouple the connection code from the download logic.

The change might not make a ton of sense in isolation, since there is no major functional change, however, decoupling the orphan handling independently helps make the review process easier and alleviates what would otherwise result in additional intermediate code to handle cases that ultimately will no longer exist.

The following is a high level overview of the changes:

- Introduce blockchain function to more easily determine if an error is a rule error with a given error code
- Move core orphan handling code from blockchain to block manager
  - Move data structures used to cache and track orphan blocks
  - Move all functions releated to orphans
    - `BlockChain.IsKnownOrphan` -> `blockManager.isKnownOrphan`
    - `BlockChain.GetOrphanRoot` -> `blockManager.orphanRoot`
    - `BlockChain.removeOrphanBlock` -> `blockManager.removeOrphanBlock`
    - `BlockChain.addOrphanBlock` -> `blockManager.addOrphanBlock`
- Implement orphan handling in block manager
  - Rework to use the moved functions and data structs
  - Add check for known orphans in addition `HaveBlock` calls to retain the same behavior
- Remove remaining orphan related code from blockchain
  - Update `ProcessBlock` to return an error when called with an orphan
  - Remove additional orphan processing from `ProcessBlock`
  - Remove orphan cache check from `HaveBlock`
  - Adjust example to account for the removed parameter
  - Change `chaingen` harness tests to detect orphans via returned error
  - Modify fullblock tests to detect orphans via returned error
  - Adapt process order logic tests to cope with lack of orphan processing
  - Update all other tests accordingly
  - Update various comments and the `README.md` and `doc.go` to properly reflect the removal of orphan handling

This is work towards #1145.

### CODE REVIEW

File: blockmanager.go [hunk 1]
```diff
@@ -1451,8 +1688,12 @@ out:
...
 
 			case processBlockMsg:
-				forkLen, isOrphan, err := b.cfg.Chain.ProcessBlock(
-					msg.block, msg.flags)
+				var isOrphan bool
+				forkLen, err := b.cfg.Chain.ProcessBlock(msg.block, msg.flags)
+				if blockchain.IsErrorCode(err, blockchain.ErrMissingParent) {
+					isOrphan = true
+					err = nil
+				}
```
Comment (matheusd) [inline blockmanager.go]: Was testing this out in simnet and I believe you need to `processOrphanBlocks`

Comment (davecgh) [reply blockmanager.go]: True.  I was considering doing that and the primary reason I decided not to was that orphans are rejected anyway if they are submitted via `getwork` or the CPU miner, but it's a good point that it's useful to retain for `submitblock`.  I'll get it updated here in a bit.

Comment (davecgh) [reply blockmanager.go]: Update to refactor the process block and orphan handling logic in block manager to a separate function that both places call.

Comment (davecgh) [reply blockmanager.go]: Output of testing the scenario in question with simnet:
```
2019-11-22 12:47:42.167 [INF] BMGR: Adding orphan block 02e1e08259ca56d43409cb1fab89861813d3fe7c1727302fc20e507ac6df8bf8 with parent 01389f17afc276b3c965c6f11c710879b20e17ea90e83d1927acab1bec6ab644
2019-11-22 12:47:42.167 [INF] RPCS: Accepted block 02e1e08259ca56d43409cb1fab89861813d3fe7c1727302fc20e507ac6df8bf8 via submitblock
2019-11-22 12:48:06.589 [INF] RPCS: Accepted block 01389f17afc276b3c965c6f11c710879b20e17ea90e83d1927acab1bec6ab644 via submitblock
```
```
$ ./ctl getbestblock
{
  "hash": "02e1e08259ca56d43409cb1fab89861813d3fe7c1727302fc20e507ac6df8bf8",
  "height": 34
}
```

File: blockmanager.go [hunk 2]
```diff
@@ -320,6 +332,13 @@ type blockManager struct {
 	startHeader      *list.Element
 	nextCheckpoint   *chaincfg.Checkpoint
 
+	// These fields are related to handling of orphan blocks.  They are
+	// protected by a combination of the chain lock and the orphan lock.
```
Comment (dnldd) [inline blockmanager.go]: Aren't the orphan related fields only protected by the orphan lock at this point?

Comment (davecgh) [reply blockmanager.go]: Yes, you're correct.  I copied the comment over and didn't update it.  Updated.

**Pull request merged**